File: 518-coin-change-ii/coin-change-ii.cpp
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        int n = coins.size();
        vector<vector<int>>dp(amount+1,vector<int>(n,0));
        for(int i=0;i<n;i++){
            dp[0][i]=1;
        }
        

        for(int i=1;i<=amount;i++){
            for(int j=n-1;j>=0;j--){
                dp[i][j] = 0;
                for(int k = j;k<n;k++){
                if(coins[k]<=i){
                   dp[i][j] += 1LL*dp[i-coins[k]][k];
                }

                }
            }
        }
        return dp[amount][0];
    }
};

File: 518-coin-change-ii/coin-change-ii.java
class Solution {
    public int change(int amount, int[] coins) {
       int n = coins.length;
       int[] a=new int[amount+1];
       for(int i=0;i<=amount;i++){
        a[i]=0;
       }
       a[0]=1;
       for(int i:coins){
        for(int j=i;j<=amount;j++){
            if(j-i>=0){a[j]+=a[j-i];}
        }
       }

       return a[amount];
    }
}